File: sdk/runanywhere-swift/Sources/RunAnywhere/Infrastructure/Events/CommonEvents.swift
//
//  CommonEvents.swift
//  RunAnywhere SDK
//
//  Common SDK events (initialization, storage, device, network).
//

import Foundation

// MARK: - SDK Lifecycle Event

/// SDK initialization and lifecycle events.
public enum SDKLifecycleEvent: SDKEvent {
    case initStarted
    case initCompleted(durationMs: Double)
    case initFailed(error: SDKError)
    case modelsLoaded(count: Int)

    public var type: String {
        switch self {
        case .initStarted: return "sdk_init_started"
        case .initCompleted: return "sdk_init_completed"
        case .initFailed: return "sdk_init_failed"
        case .modelsLoaded: return "sdk_models_loaded"
        }
    }

    public var category: EventCategory { .sdk }

    public var properties: [String: String] {
        switch self {
        case .initStarted:
            return [:]
        case .initCompleted(let durationMs):
            return ["duration_ms": String(format: "%.1f", durationMs)]
        case .initFailed(let error):
            return error.telemetryProperties
        case .modelsLoaded(let count):
            return ["count": String(count)]
        }
    }
}

// MARK: - Model Event

/// Common model lifecycle events (download, delete).
/// For capability-specific load/unload, use LLMEvent, STTEvent, TTSEvent.
public enum ModelEvent: SDKEvent {
    // Download events
    case downloadStarted(modelId: String)
    case downloadProgress(modelId: String, progress: Double, bytesDownloaded: Int64, totalBytes: Int64)
    case downloadCompleted(modelId: String, durationMs: Double, sizeBytes: Int64)
    case downloadFailed(modelId: String, error: SDKError)
    case downloadCancelled(modelId: String)

    // Extraction events (for archive-based models)
    case extractionStarted(modelId: String, archiveType: String)
    case extractionProgress(modelId: String, progress: Double)
    case extractionCompleted(modelId: String, durationMs: Double)
    case extractionFailed(modelId: String, error: SDKError)

    // Deletion events
    case deleted(modelId: String)

    public var type: String {
        switch self {
        case .downloadStarted: return "model_download_started"
        case .downloadProgress: return "model_download_progress"
        case .downloadCompleted: return "model_download_completed"
        case .downloadFailed: return "model_download_failed"
        case .downloadCancelled: return "model_download_cancelled"
        case .extractionStarted: return "model_extraction_started"
        case .extractionProgress: return "model_extraction_progress"
        case .extractionCompleted: return "model_extraction_completed"
        case .extractionFailed: return "model_extraction_failed"
        case .deleted: return "model_deleted"
        }
    }

    public var category: EventCategory { .model }

    public var destination: EventDestination {
        switch self {
        case .downloadProgress, .extractionProgress:
            // Progress updates are for public EventBus (UI updates) - not sent to analytics/remote
            return .publicOnly
        default:
            return .all
        }
    }

    public var properties: [String: String] {
        switch self {
        case .downloadStarted(let modelId):
            return ["model_id": modelId]

        case .downloadProgress(let modelId, let progress, let bytes, let total):
            return [
                "model_id": modelId,
                "progress": String(format: "%.1f", progress),
                "bytes_downloaded": String(bytes),
                "total_bytes": String(total)
            ]

        case .downloadCompleted(let modelId, let durationMs, let sizeBytes):
            return [
                "model_id": modelId,
                "duration_ms": String(format: "%.1f", durationMs),
                "size_bytes": String(sizeBytes)
            ]

        case .downloadFailed(let modelId, let error):
            return ["model_id": modelId].merging(error.telemetryProperties) { _, new in new }

        case .downloadCancelled(let modelId):
            return ["model_id": modelId]

        case .extractionStarted(let modelId, let archiveType):
            return ["model_id": modelId, "archive_type": archiveType]

        case .extractionProgress(let modelId, let progress):
            return [
                "model_id": modelId,
                "progress": String(format: "%.1f", progress)
            ]

        case .extractionCompleted(let modelId, let durationMs):
            return [
                "model_id": modelId,
                "duration_ms": String(format: "%.1f", durationMs)
            ]

        case .extractionFailed(let modelId, let error):
            return ["model_id": modelId].merging(error.telemetryProperties) { _, new in new }

        case .deleted(let modelId):
            return ["model_id": modelId]
        }
    }
}

// MARK: - Storage Event

/// Storage-related events.
public enum StorageEvent: SDKEvent {
    case cacheCleared(freedBytes: Int64)
    case cacheClearFailed(error: SDKError)
    case tempFilesCleaned(freedBytes: Int64)

    public var type: String {
        switch self {
        case .cacheCleared: return "storage_cache_cleared"
        case .cacheClearFailed: return "storage_cache_clear_failed"
        case .tempFilesCleaned: return "storage_temp_cleaned"
        }
    }

    public var category: EventCategory { .storage }

    public var properties: [String: String] {
        switch self {
        case .cacheCleared(let freedBytes):
            return ["freed_bytes": String(freedBytes)]
        case .cacheClearFailed(let error):
            return error.telemetryProperties
        case .tempFilesCleaned(let freedBytes):
            return ["freed_bytes": String(freedBytes)]
        }
    }
}

// MARK: - Device Event

/// Device-related events.
public enum DeviceEvent: SDKEvent {
    case registered(deviceId: String)
    case registrationFailed(error: SDKError)

    public var type: String {
        switch self {
        case .registered: return "device_registered"
        case .registrationFailed: return "device_registration_failed"
        }
    }

    public var category: EventCategory { .device }

    public var properties: [String: String] {
        switch self {
        case .registered(let deviceId):
            return ["device_id": deviceId]
        case .registrationFailed(let error):
            return error.telemetryProperties
        }
    }
}

// MARK: - Network Event

/// Network-related events.
public enum NetworkEvent: SDKEvent {
    case connectivityChanged(isOnline: Bool)

    public var type: String {
        switch self {
        case .connectivityChanged: return "network_connectivity_changed"
        }
    }

    public var category: EventCategory { .network }

    public var destination: EventDestination {
        .analyticsOnly
    }

    public var properties: [String: String] {
        switch self {
        case .connectivityChanged(let isOnline):
            return ["is_online": String(isOnline)]
        }
    }
}

// MARK: - SDK Error Event

/// SDK error events - exposed to consumers for error handling and logging.
/// These events carry full error context including stack traces for debugging.
///
/// Consumers can subscribe to these events via EventBus:
/// ```swift
/// EventBus.shared.events(for: .error)
///     .compactMap { $0 as? SDKErrorEvent }
///     .sink { errorEvent in
///         // Log to Sentry, Crashlytics, etc.
///         print("Error: \(errorEvent.error.code) - \(errorEvent.error.message)")
///         print("Stack: \(errorEvent.error.stackTrace)")
///     }
/// ```
public struct SDKErrorEvent: SDKEvent, TypedEventProperties {

    // MARK: - Properties

    /// The underlying SDK error with full context (code, message, category, stack trace)
    public let error: SDKError

    /// The operation that was being performed when the error occurred
    public let operation: String

    /// Additional context about the error (URL, status code, response body, etc.)
    public let context: [String: String]

    // MARK: - SDKEvent Protocol

    public let id: String
    public let timestamp: Date
    public var type: String { "sdk_error" }
    public var category: EventCategory { .error }
    public var sessionId: String? { nil }

    /// Error events go to both consumers (EventBus) and analytics
    /// Consumers can listen to these for error handling/logging
    public var destination: EventDestination { .all }

    public var properties: [String: String] {
        var props = error.telemetryProperties
        props["operation"] = operation

        // Merge context
        for (key, value) in context {
            props[key] = value
        }

        return props
    }

    // MARK: - TypedEventProperties

    public var typedProperties: EventProperties {
        EventProperties(
            errorMessage: error.message,
            errorCode: error.code.rawValue
        )
    }

    // MARK: - Initialization

    public init(
        error: SDKError,
        operation: String,
        context: [String: String] = [:]
    ) {
        self.id = UUID().uuidString
        self.timestamp = Date()
        self.error = error
        self.operation = operation
        self.context = context
    }

    // MARK: - Factory Methods

    /// Create an error event for a network failure (used by APIClient)
    public static func networkError(
        error: SDKError,
        url: String? = nil,
        statusCode: Int? = nil,
        responseBody: String? = nil
    ) -> SDKErrorEvent {
        var context: [String: String] = [:]
        if let url = url {
            context["url"] = url
        }
        if let statusCode = statusCode {
            context["status_code"] = String(statusCode)
        }
        if let body = responseBody {
            // Truncate long response bodies for logging
            context["response_body"] = String(body.prefix(500))
        }
        return SDKErrorEvent(error: error, operation: "network_request", context: context)
    }

    /// Create an error event for LLM operations (used by analytics services)
    public static func llmError(
        error: SDKError,
        modelId: String? = nil,
        generationId: String? = nil,
        operation: String
    ) -> SDKErrorEvent {
        var context: [String: String] = ["capability": "llm"]
        if let modelId = modelId {
            context["model_id"] = modelId
        }
        if let generationId = generationId {
            context["generation_id"] = generationId
        }
        return SDKErrorEvent(error: error, operation: operation, context: context)
    }

    /// Create an error event for STT operations (used by analytics services)
    public static func sttError(
        error: SDKError,
        modelId: String? = nil,
        transcriptionId: String? = nil,
        operation: String
    ) -> SDKErrorEvent {
        var context: [String: String] = ["capability": "stt"]
        if let modelId = modelId {
            context["model_id"] = modelId
        }
        if let transcriptionId = transcriptionId {
            context["transcription_id"] = transcriptionId
        }
        return SDKErrorEvent(error: error, operation: operation, context: context)
    }

    /// Create an error event for TTS operations (used by analytics services)
    public static func ttsError(
        error: SDKError,
        modelId: String? = nil,
        synthesisId: String? = nil,
        operation: String
    ) -> SDKErrorEvent {
        var context: [String: String] = ["capability": "tts"]
        if let modelId = modelId {
            context["model_id"] = modelId
        }
        if let synthesisId = synthesisId {
            context["synthesis_id"] = synthesisId
        }
        return SDKErrorEvent(error: error, operation: operation, context: context)
    }

    /// Create from any Error, converting to SDKError if needed
    public static func from(
        _ error: Error,
        operation: String,
        context: [String: String] = [:]
    ) -> SDKErrorEvent {
        let sdkError = SDKError.from(error)
        return SDKErrorEvent(error: sdkError, operation: operation, context: context)
    }
}

// MARK: - Framework Event

/// Framework query events (analytics only).
public enum FrameworkEvent: SDKEvent {
    case modelsRequested(framework: String)
    case modelsRetrieved(framework: String, count: Int)

    public var type: String {
        switch self {
        case .modelsRequested: return "framework_models_requested"
        case .modelsRetrieved: return "framework_models_retrieved"
        }
    }

    public var category: EventCategory { .sdk }
    public var destination: EventDestination { .analyticsOnly }

    public var properties: [String: String] {
        switch self {
        case .modelsRequested(let framework):
            return ["framework": framework]
        case .modelsRetrieved(let framework, let count):
            return ["framework": framework, "count": String(count)]
        }
    }
}

// MARK: - Voice Pipeline Event

/// Voice pipeline events (VAD, full pipeline).
public enum VoicePipelineEvent: SDKEvent {
    case pipelineStarted
    case pipelineCompleted(durationMs: Double)
    case pipelineFailed(error: SDKError)
    case vadStarted
    case speechDetected
    case speechEnded
    case vadStopped

    public var type: String {
        switch self {
        case .pipelineStarted: return "voice_pipeline_started"
        case .pipelineCompleted: return "voice_pipeline_completed"
        case .pipelineFailed: return "voice_pipeline_failed"
        case .vadStarted: return "vad_started"
        case .speechDetected: return "vad_speech_detected"
        case .speechEnded: return "vad_speech_ended"
        case .vadStopped: return "vad_stopped"
        }
    }

    public var category: EventCategory { .voice }

    public var properties: [String: String] {
        switch self {
        case .pipelineStarted, .vadStarted, .speechDetected, .speechEnded, .vadStopped:
            return [:]
        case .pipelineCompleted(let durationMs):
            return ["duration_ms": String(format: "%.1f", durationMs)]
        case .pipelineFailed(let error):
            return error.telemetryProperties
        }
    }
}

File: sdk/runanywhere-react-native/packages/core/src/Infrastructure/Events/CommonEvents.ts
/**
 * CommonEvents
 *
 * SDK-wide event definitions for the RunAnywhere SDK.
 * These are common events used across the SDK (initialization, storage, device, network).
 *
 * For capability-specific events (LLM, STT, TTS), see the respective capability modules.
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Infrastructure/Events/CommonEvents.swift
 */

import type { SDKEvent } from './SDKEvent';
import { EventCategory, EventDestination, createSDKEvent } from './SDKEvent';

// ============================================================================
// SDK Lifecycle Events
// ============================================================================

/**
 * SDK Lifecycle Event Types
 */
export type SDKLifecycleEventType =
  | 'sdk_init_started'
  | 'sdk_init_completed'
  | 'sdk_init_failed'
  | 'sdk_models_loaded';

/**
 * Create SDK initialization started event.
 */
export function createSDKInitStartedEvent(sessionId?: string): SDKEvent {
  return createSDKEvent(
    'sdk_init_started',
    EventCategory.SDK,
    {},
    { sessionId }
  );
}

/**
 * Create SDK initialization completed event.
 * @param durationMs - Time taken to initialize in milliseconds
 */
export function createSDKInitCompletedEvent(
  durationMs: number,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'sdk_init_completed',
    EventCategory.SDK,
    { duration_ms: durationMs.toFixed(1) },
    { sessionId }
  );
}

/**
 * Create SDK initialization failed event.
 * @param error - Error message
 */
export function createSDKInitFailedEvent(
  error: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'sdk_init_failed',
    EventCategory.SDK,
    { error },
    { sessionId }
  );
}

/**
 * Create SDK models loaded event.
 * @param count - Number of models loaded
 */
export function createSDKModelsLoadedEvent(
  count: number,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'sdk_models_loaded',
    EventCategory.SDK,
    { count: String(count) },
    { sessionId }
  );
}

// ============================================================================
// Model Events
// ============================================================================

/**
 * Model Event Types
 *
 * Common model lifecycle events (download, extraction, delete).
 * For capability-specific load/unload, use LLMEvent, STTEvent, TTSEvent.
 */
export type ModelEventType =
  | 'model_download_started'
  | 'model_download_progress'
  | 'model_download_completed'
  | 'model_download_failed'
  | 'model_download_cancelled'
  | 'model_extraction_started'
  | 'model_extraction_progress'
  | 'model_extraction_completed'
  | 'model_extraction_failed'
  | 'model_deleted';

/**
 * Create model download started event.
 */
export function createModelDownloadStartedEvent(
  modelId: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'model_download_started',
    EventCategory.Model,
    { model_id: modelId },
    { sessionId }
  );
}

/**
 * Create model download progress event.
 * Note: Progress events are analytics-only (too chatty for public EventBus).
 */
export function createModelDownloadProgressEvent(
  modelId: string,
  progress: number,
  bytesDownloaded: number,
  totalBytes: number,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'model_download_progress',
    EventCategory.Model,
    {
      model_id: modelId,
      progress: progress.toFixed(1),
      bytes_downloaded: String(bytesDownloaded),
      total_bytes: String(totalBytes),
    },
    { sessionId, destination: EventDestination.AnalyticsOnly }
  );
}

/**
 * Create model download completed event.
 */
export function createModelDownloadCompletedEvent(
  modelId: string,
  durationMs: number,
  sizeBytes: number,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'model_download_completed',
    EventCategory.Model,
    {
      model_id: modelId,
      duration_ms: durationMs.toFixed(1),
      size_bytes: String(sizeBytes),
    },
    { sessionId }
  );
}

/**
 * Create model download failed event.
 */
export function createModelDownloadFailedEvent(
  modelId: string,
  error: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'model_download_failed',
    EventCategory.Model,
    { model_id: modelId, error },
    { sessionId }
  );
}

/**
 * Create model download cancelled event.
 */
export function createModelDownloadCancelledEvent(
  modelId: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'model_download_cancelled',
    EventCategory.Model,
    { model_id: modelId },
    { sessionId }
  );
}

/**
 * Create model extraction started event.
 */
export function createModelExtractionStartedEvent(
  modelId: string,
  archiveType: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'model_extraction_started',
    EventCategory.Model,
    { model_id: modelId, archive_type: archiveType },
    { sessionId }
  );
}

/**
 * Create model extraction progress event.
 * Note: Progress events are analytics-only (too chatty for public EventBus).
 */
export function createModelExtractionProgressEvent(
  modelId: string,
  progress: number,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'model_extraction_progress',
    EventCategory.Model,
    { model_id: modelId, progress: progress.toFixed(1) },
    { sessionId, destination: EventDestination.AnalyticsOnly }
  );
}

/**
 * Create model extraction completed event.
 */
export function createModelExtractionCompletedEvent(
  modelId: string,
  durationMs: number,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'model_extraction_completed',
    EventCategory.Model,
    { model_id: modelId, duration_ms: durationMs.toFixed(1) },
    { sessionId }
  );
}

/**
 * Create model extraction failed event.
 */
export function createModelExtractionFailedEvent(
  modelId: string,
  error: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'model_extraction_failed',
    EventCategory.Model,
    { model_id: modelId, error },
    { sessionId }
  );
}

/**
 * Create model deleted event.
 */
export function createModelDeletedEvent(
  modelId: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'model_deleted',
    EventCategory.Model,
    { model_id: modelId },
    { sessionId }
  );
}

// ============================================================================
// Storage Events
// ============================================================================

/**
 * Storage Event Types
 */
export type StorageEventType =
  | 'storage_cache_cleared'
  | 'storage_cache_clear_failed'
  | 'storage_temp_cleaned';

/**
 * Create storage cache cleared event.
 */
export function createStorageCacheClearedEvent(
  freedBytes: number,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'storage_cache_cleared',
    EventCategory.Storage,
    { freed_bytes: String(freedBytes) },
    { sessionId }
  );
}

/**
 * Create storage cache clear failed event.
 */
export function createStorageCacheClearFailedEvent(
  error: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'storage_cache_clear_failed',
    EventCategory.Storage,
    { error },
    { sessionId }
  );
}

/**
 * Create temp files cleaned event.
 */
export function createStorageTempFilesCleanedEvent(
  freedBytes: number,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'storage_temp_cleaned',
    EventCategory.Storage,
    { freed_bytes: String(freedBytes) },
    { sessionId }
  );
}

// ============================================================================
// Device Events
// ============================================================================

/**
 * Device Event Types
 */
export type DeviceEventType =
  | 'device_registered'
  | 'device_registration_failed';

/**
 * Create device registered event.
 */
export function createDeviceRegisteredEvent(
  deviceId: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'device_registered',
    EventCategory.Device,
    { device_id: deviceId },
    { sessionId }
  );
}

/**
 * Create device registration failed event.
 */
export function createDeviceRegistrationFailedEvent(
  error: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'device_registration_failed',
    EventCategory.Device,
    { error },
    { sessionId }
  );
}

// ============================================================================
// Network Events
// ============================================================================

/**
 * Network Event Types
 */
export type NetworkEventType =
  | 'network_connectivity_changed'
  | 'network_request_failed';

/**
 * Create network connectivity changed event.
 * Note: Network events are analytics-only by default.
 */
export function createNetworkConnectivityChangedEvent(
  isOnline: boolean,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'network_connectivity_changed',
    EventCategory.Network,
    { is_online: String(isOnline) },
    { sessionId, destination: EventDestination.AnalyticsOnly }
  );
}

/**
 * Create network request failed event.
 * Note: Network events are analytics-only by default.
 */
export function createNetworkRequestFailedEvent(
  url: string,
  error: string,
  statusCode?: number,
  sessionId?: string
): SDKEvent {
  const properties: Record<string, string> = { url, error };
  if (statusCode !== undefined) {
    properties.status_code = String(statusCode);
  }

  return createSDKEvent(
    'network_request_failed',
    EventCategory.Network,
    properties,
    {
      sessionId,
      destination: EventDestination.AnalyticsOnly,
    }
  );
}

// ============================================================================
// Error Events
// ============================================================================

/**
 * Error Event Types
 */
export type ErrorEventType = 'error';

/**
 * Create general error event.
 * Note: Error events are analytics-only by default.
 */
export function createErrorEvent(
  operation: string,
  message: string,
  code?: number,
  sessionId?: string
): SDKEvent {
  const properties: Record<string, string> = { operation, message };
  if (code !== undefined) {
    properties.code = String(code);
  }

  return createSDKEvent('error', EventCategory.Error, properties, {
    sessionId,
    destination: EventDestination.AnalyticsOnly,
  });
}

// ============================================================================
// Framework Events
// ============================================================================

/**
 * Framework Event Types
 */
export type FrameworkEventType =
  | 'framework_models_requested'
  | 'framework_models_retrieved';

/**
 * Create framework models requested event.
 * Note: Framework events are analytics-only.
 */
export function createFrameworkModelsRequestedEvent(
  framework: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'framework_models_requested',
    EventCategory.SDK,
    { framework },
    { sessionId, destination: EventDestination.AnalyticsOnly }
  );
}

/**
 * Create framework models retrieved event.
 * Note: Framework events are analytics-only.
 */
export function createFrameworkModelsRetrievedEvent(
  framework: string,
  count: number,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'framework_models_retrieved',
    EventCategory.SDK,
    { framework, count: String(count) },
    { sessionId, destination: EventDestination.AnalyticsOnly }
  );
}

// ============================================================================
// Voice Pipeline Events
// ============================================================================

/**
 * Voice Pipeline Event Types
 */
export type VoicePipelineEventType =
  | 'voice_pipeline_started'
  | 'voice_pipeline_completed'
  | 'voice_pipeline_failed'
  | 'vad_started'
  | 'vad_speech_detected'
  | 'vad_speech_ended'
  | 'vad_stopped';

/**
 * Create voice pipeline started event.
 */
export function createVoicePipelineStartedEvent(sessionId?: string): SDKEvent {
  return createSDKEvent(
    'voice_pipeline_started',
    EventCategory.Voice,
    {},
    { sessionId }
  );
}

/**
 * Create voice pipeline completed event.
 */
export function createVoicePipelineCompletedEvent(
  durationMs: number,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'voice_pipeline_completed',
    EventCategory.Voice,
    { duration_ms: durationMs.toFixed(1) },
    { sessionId }
  );
}

/**
 * Create voice pipeline failed event.
 */
export function createVoicePipelineFailedEvent(
  error: string,
  sessionId?: string
): SDKEvent {
  return createSDKEvent(
    'voice_pipeline_failed',
    EventCategory.Voice,
    { error },
    { sessionId }
  );
}

/**
 * Create VAD started event.
 */
export function createVADStartedEvent(sessionId?: string): SDKEvent {
  return createSDKEvent('vad_started', EventCategory.Voice, {}, { sessionId });
}

/**
 * Create speech detected event.
 */
export function createSpeechDetectedEvent(sessionId?: string): SDKEvent {
  return createSDKEvent(
    'vad_speech_detected',
    EventCategory.Voice,
    {},
    { sessionId }
  );
}

/**
 * Create speech ended event.
 */
export function createSpeechEndedEvent(sessionId?: string): SDKEvent {
  return createSDKEvent(
    'vad_speech_ended',
    EventCategory.Voice,
    {},
    { sessionId }
  );
}

/**
 * Create VAD stopped event.
 */
export function createVADStoppedEvent(sessionId?: string): SDKEvent {
  return createSDKEvent('vad_stopped', EventCategory.Voice, {}, { sessionId });
}

// ============================================================================
// Combined Event Types (for type guards and filtering)
// ============================================================================

/**
 * All common event types.
 */
export type CommonEventType =
  | SDKLifecycleEventType
  | ModelEventType
  | StorageEventType
  | DeviceEventType
  | NetworkEventType
  | ErrorEventType
  | FrameworkEventType
  | VoicePipelineEventType;
